src/engine/Result.cpp

```diff
@@ -216,39 +231,27 @@ void Result::runOnNewChunkComputed(
         onNewChunk,
     std::function<void(bool)> onGeneratorFinished) {
   AD_CONTRACT_CHECK(!isFullyMaterialized());
-  auto generator = [](LazyResult original, auto onNewChunk,
-                      auto onGeneratorFinished) -> Generator {
-    // Call this within destructor to make sure it is also called when an
-    // operation stops iterating before reaching the end.
-    absl::Cleanup cleanup{
-        [&onGeneratorFinished]() { onGeneratorFinished(false); }};
-    try {
-      ad_utility::timer::Timer timer{ad_utility::timer::Timer::Started};
-      for (IdTableVocabPair& pair : original) {
-        onNewChunk(pair, timer.value());
-        co_yield pair;
-        timer.start();
-      }
-    } catch (...) {
-      std::move(cleanup).Cancel();
-      onGeneratorFinished(true);
-      throw;
-    }
-  }(std::move(idTables()), std::move(onNewChunk),
-                                                std::move(onGeneratorFinished));
-  data_.emplace<GenContainer>(std::move(generator));
-}
-
-// _____________________________________________________________________________
-void Result::assertSortOrderIsRespected(
-    const IdTable& idTable, const std::vector<ColumnIndex>& sortedBy) {
-  AD_CONTRACT_CHECK(
-      ql::ranges::all_of(sortedBy, [&idTable](ColumnIndex colIndex) {
-        return colIndex < idTable.numColumns();
-      }));
-
-  AD_EXPENSIVE_CHECK(
-      ql::ranges::is_sorted(idTable, compareRowsBySortColumns(sortedBy)));
+  ad_utility::CachingTransformInputRange generator{
+      std::move(idTables()),
+      [onNewChunk = std::move(onNewChunk), &onGeneratorFinished,
+       // Call this within destructor to make sure it is also called when an
+       // operation stops iterating before reaching the end.
+       cleanup = absl::Cleanup{[&onGeneratorFinished]() {
+         onGeneratorFinished(false);
```

Comment:
For this generator please read my EMail.
Also currently the `onGeneratorFinished` is captured by reference which might unfortunately be dangling.
I checked: It is not possible to move-capture it and then capture that again in the other capture (the cleanup capture), but you could move the `onGeneratorFinished` into a `shared_ptr` or `unique_ptr` and then think of some way to capture it without too many copies.